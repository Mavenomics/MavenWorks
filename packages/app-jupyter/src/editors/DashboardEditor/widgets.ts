import { PartManager } from "@mavenomics/dashboard";
import { KernelProxyPart } from "@mavenomics/jupyterutils";
import { WidgetRenderer, registerWidgetManager } from "@jupyter-widgets/jupyterlab-manager";
import { JupyterFrontEndPlugin } from "@jupyterlab/application";
import { IDashboardEditorTracker } from "./plugin";
import { IJupyterWidgetRegistry } from "@jupyter-widgets/base";

/** Helper plugin that adds IPyWidget support to Dashboard Docs. */

function* collectWidgetRenderers(partManager: PartManager) {
    for (const [_, part] of partManager) {
        if (!(part instanceof KernelProxyPart)) continue;
        if (part.renderer instanceof WidgetRenderer) {
            yield part.renderer;
        }
    }
}

const plugin: JupyterFrontEndPlugin<void> = {
    id: "jupyterlab-mavenworks:dashboard-doc-widget-manager",
    autoStart: true,
    requires: [
        IDashboardEditorTracker
    ],
    optional: [
        // if the widgets aren't there do nothing
        IJupyterWidgetRegistry
    ],
    activate: (
        _app,
        tracker: IDashboardEditorTracker,
        registry?: IJupyterWidgetRegistry
    ) => {
        if (registry == null) return; // ipywidgets isn't installed

        tracker.forEach(i => {
            registerWidgetManager(
                i.context as any,
                i.rendermime,
                collectWidgetRenderers(i.content.partManager)
            );
        });

        tracker.widgetAdded.connect((_, i) => {
            registerWidgetManager(
                i.context as any,
                i.rendermime,
                collectWidgetRenderers(i.content.partManager)
            );
        });
    }
};
export default plugin;
