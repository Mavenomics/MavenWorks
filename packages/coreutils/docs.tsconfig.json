{
    "extends": "./tsconfig.json",
    "compilerOptions": {
        "outDir": "./lib",
        "rootDir": "./src"
    },
    "files": [
        "./src/index.ts"
    ],
    "typedocOptions": {
        "mode": "file",
        "out": "../../docs/api/coreutils"
    }
}