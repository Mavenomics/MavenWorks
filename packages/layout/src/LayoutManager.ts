import { Widget, PanelLayout, StackedLayout } from "@phosphor/widgets";
import { StackPanelLayoutRegion } from "./StackPanelLayoutRegion";
import { LayoutSerializer } from "./LayoutSerializer";
import { RegionWithChildren } from "./RegionWithChildren";
import { DashboardLayoutRegion } from "./DashboardLayoutRegion";
import { MessageLoop, Message } from "@phosphor/messaging";
import { Signal, ISignal } from "@phosphor/signaling";
import { Subject } from "rxjs";
import { LayoutActions } from "./layoutactions";
import { DockPreview } from "./dockpreview";
import { IterTools, IDirtyable } from "@mavenomics/coreutils";
import { ReactWrapperWidget, HoverManager, ListBox } from "@mavenomics/ui";
import * as React from "react";
import { WidgetLayoutRegion } from "./WidgetLayoutRegion";

/**
 * Enum for layout type names
 */
export enum LayoutTypes {
    StackPanelLayoutRegion,
    WidgetLayoutRegion,
    TabPanelDashboardLayoutRegion,
    CanvasLayoutRegion,
    GridPanelLayoutRegion
}

/**
 * Base layout for Maven Dashboards
 *
 * This class manages the layout operations of a set of widgets, which are
 * arranged in a tree and consist of DashboardLayoutRegions. These regions have
 * a set of layout properties that describe appearance and layout behavior- for
 * instance, all regions have properties like `borderWidth.px` and
 * `borderColor`.
 */
export class LayoutManager extends Widget implements IDirtyable {
    public layout: PanelLayout;
    public dockPreview: DockPreview;
    public root: StackPanelLayoutRegion;
    private partManager: LayoutManager.IPartManager;
    private factory: LayoutManager.IFactory;
    private externalParts = new Map<string, Widget>();
    private _maximizedRegion: DashboardLayoutRegion<any> | null = null;
    private _focusedRegion: DashboardLayoutRegion<any> | null = null;
    private _onFocusedChanged = new Signal<this, DashboardLayoutRegion<any> | null>(this);
    private _isDirty = false;
    private _OnDirtySrc = new Subject<void>();
    private _OnDirty = this._OnDirtySrc.asObservable();
    private _forceTitlebars = false;

    constructor(partManager: LayoutManager.IPartManager, factory: LayoutManager.IFactory) {
        super();
        this.node.tabIndex = -1;
        this.addClass("m-LayoutManager");
        this.partManager = partManager;
        this.factory = factory;
        this.layout = new StackedLayout();
        this.root = new StackPanelLayoutRegion(this);
        this.root.setLayoutProperty("caption", "Dashboard");
        this.root.setFresh();
        this.root.OnStale.subscribe(() => this.setDirty());
        this.dockPreview = new DockPreview({ owner: this });
        this.dockPreview.hide();
        this.layout.addWidget(this.root);
        this.layout.addWidget(this.dockPreview);
    }

    /** The layout region that currently has focus, if any. */
    public get focusedRegion() { return this._focusedRegion; }

    /**
     * A signal that emits when the focused layout region changes.
     *
     * #### Notes
     *
     * A region's focus is not related to DOM focus.
     *
     * @see DashboardLayoutRegion.LayoutMsg.BeforeFocus*/
    public get focusedRegionChanged(): ISignal<this, DashboardLayoutRegion<any> | null> {
        return this._onFocusedChanged;
    }

    /** Whether this layout has unsaved changes */
    public get isDirty() { return this._isDirty; }

    /** An Observable that emits whenever this layout becomes dirty */
    public get OnDirty() { return this._OnDirty; }

    /** If true, override [[showTitle]] on all layout regions with `true`. */
    public get forceTitlebars() { return this._forceTitlebars; }

    /** Set forceTitlebars and update the layout */
    public set forceTitlebars(force: boolean) {
        if (force === this._forceTitlebars) return;
        this._forceTitlebars = force;
        this.setDirty();
    }

    /** Mark this layout as having been synchronized with the model */
    public setClean() {
        this._isDirty = false;
        this.update();
    }

    // TODO: Remove this
    getPart(id: string) {
        return this.partManager.getPartById(id) || this.externalParts.get(id);
    }

    setExternalParts(parts: Map<string, Widget>) {
        this.externalParts = parts;
    }

    initLayout(serializedLayout: LayoutSerializer.ISerializedLayoutRegion) {
        if (!serializedLayout) {
            return; // no layout to setup
        }

        this.root.dispose();
        this.root = LayoutSerializer.fromJson(serializedLayout, this) as StackPanelLayoutRegion;
        this.root.setFresh();
        this.root.OnStale.subscribe(() => this.setDirty());
        this.root.titlebar.hide(); // TODO: RootLayoutRegion?
        this.layout.insertWidget(0, this.root);
    }

    public getRegion(uuid: string): DashboardLayoutRegion<any> | null {
        // TODO: map for storing region UUIDs
        if (uuid === this.root.uuid) {
            return this.root;
        }
        for (const region of IterTools.dfs_iter(
            this.root.widgets,
            i => i instanceof RegionWithChildren ? i.widgets : undefined
        )) {
            if (region.uuid === uuid) {
                return region;
            }
        }
        return null; // region not found
    }

    public getParentRegion(uuid: string) {
        // TODO: map for storing region UUIDs
        if (uuid === this.root.uuid) {
            return this.root; // root is it's own parent
        }
        const region = this.getRegion(uuid);
        if (region != null && region.parentRegion != null) {
            return region.parentRegion;
        }
        return null; // parent not found
    }

    public removeRegion(uuid: string) {
        const toRemove = this.getRegion(uuid);
        if (!toRemove) {
            throw Error("Region with " + uuid + " not found!");
        }
        const parent = toRemove.parentRegion;
        toRemove.dispose();
        this.setDirty();
        if (parent) {
            parent.fit();
            parent.update();
        }
        this.root.pruneSubtree();
    }

    public dispose() {
        if (this.isDisposed) {
            return;
        }
        this._OnDirtySrc.complete();
        this._focusedRegion = null;
        this.root.dispose();
        this.externalParts.forEach(i => i.dispose());
        this.externalParts.clear();
        super.dispose();
    }

    public handleEvent(ev: FocusEvent): void {
        switch (ev.type) {
            case "focusin":
            case "mousedown":
                return this._handleFocus(ev);
        }
    }

    public moveToRootZone(zone: LayoutActions.RootDropZone, target: string) {
        return LayoutActions.MoveToRootZone({
            layoutManager: this,
            zone,
            target
        });
    }

    public moveToRelativeZone(
        zone: LayoutActions.RelativeDropZone,
        target: string,
        reference: string
    ) {
        return LayoutActions.MoveToRelativeZone({
            layoutManager: this,
            zone,
            target,
            reference
        });
    }

    public async showAddPartDialog() {
        const items: ListBox.ListItem[] = [];
        for (const name of this.factory.keys()) {
            items.push({
                key: name,
                label: name
            });
        }
        items.sort((a, b) => a.key.localeCompare(b.key));
        let selectedPart: string | null = null;
        const hover = ReactWrapperWidget.Create(
            () => React.createElement<ListBox.IProps>(ListBox, {
                items,
                selectedKey: selectedPart,
                onSelect: (key) => {
                    selectedPart = key;
                    hover.update();
                },
                onCommit: (key) => {
                    selectedPart = key;
                    MessageLoop.sendMessage(
                        hover.parent!,
                        HoverManager.DialogMsg.AcceptRequest
                    );
                },
                isEditing: false,
                onEdit: () => void 0,
            })
        );
        hover.node.style.overflowY = "auto";
        const res = await HoverManager.Instance!.launchDialog(
            hover,
            this,
            300,
            600,
            "Add New Part",
            [{ text: "Dismiss" }, { text: "Ok", accept: true }]
        );
        if (!res.accept) return;
        return selectedPart;
    }

    public async addPartToRegion(target: string, type: string) {
        const region = this.getRegion(target);
        if (!(region instanceof RegionWithChildren)) {
            throw Error("Cannot add region to parent");
        }
        const widget = await this.factory.get(type);
        const widgetRegion = new WidgetLayoutRegion(this, widget, widget.id);
        region.addChild(widgetRegion);
    }

    get bounds(): { width: number, height: number } {
        const bounds = (this.node.getBoundingClientRect());
        return { width: bounds.width, height: bounds.height };
    }

    protected updateMaximizedRegion() {
        let regionToMaximize = null;
        for (const region of IterTools.dfs_iter(
            this.root.widgets,
            i => i instanceof RegionWithChildren ? i.widgets : undefined
        )) {
            //Ignore the currently maximized region.
            //This allows us to handle the case where someone maximizes a region inside the designer
            if (this._maximizedRegion === region)
                continue;

            if (region.getLayoutProperty("maximized")) {
                if (!regionToMaximize) {
                    regionToMaximize = region;
                } else {
                    //Only allow 1 maximized region
                    region.setLayoutProperty("maximized", false);
                }
            }
        }

        //Handle the case where the root is maximized via the designer
        if (this.root.getLayoutProperty("maximized"))
            this.root.setLayoutProperty("maximized", false);

        //Nothing is maximized
        if (!this._maximizedRegion && !regionToMaximize)
            return;

        //Maximized region is still maximized and there is nothing new to maximize
        if (!regionToMaximize && this._maximizedRegion && this._maximizedRegion.getLayoutProperty("maximized"))
            return;

        //Remove the already maximized widget.
        //This is hit when either replacing the maximized widget or unmaximizing the widget
        if (this._maximizedRegion != null) {
            this._maximizedRegion.setLayoutProperty("maximized", false);
            //this.layout.removeWidget(this._maximizedRegion);
            document.querySelectorAll(".no-stack").forEach(i => i.classList.remove("no-stack"));
            this._maximizedRegion = null;
        }

        //Maximize the new widget
        if (regionToMaximize != null) {
            regionToMaximize.setLayoutProperty("maximized", true);
            this._maximizedRegion = regionToMaximize;
            let cur = regionToMaximize.parent;
            while (cur && cur !== this.root) {
                cur.addClass("no-stack");
                cur = cur.parent;
            }
            //this.layout.addWidget(regionToMaximize);
            this.setDirty();
        }
    }

    protected onUpdateRequest() {
        this.updateMaximizedRegion();
    }

    protected onAfterAttach() {
        this.node.addEventListener("focusin", this);
        this.node.addEventListener("mousedown", this);
    }

    protected onBeforeDetach() {
        this.node.removeEventListener("focusin", this);
        this.node.removeEventListener("mousedown", this);
    }

    private _handleFocus(ev: FocusEvent) {
        let target = ev.target as HTMLElement;
        // We can do this check because we won't get events *directly* from
        // iframes in parts. Instead, they'll be bubbled up via the event
        // copying mechanism.
        if (!this.root.node.contains(target)) {
            return; // not something we can focus
        }
        while (target.parentElement != null && target !== this.node) {
            if (!target.classList.contains("maven_dashboard")) {
                target = target.parentElement;
                continue;
            }
            const regionId = target.id;
            const region = this.getRegion(regionId);
            if (region == null) {
                return; // region isn't valid
                // though if this happens, something else is also wrong
            }
            this.setFocusedRegion(region);
            return;
        }
    }

    private setFocusedRegion(region: DashboardLayoutRegion<any>) {
        if (region === this._focusedRegion) {
            // do nothing
            return;
        }
        const oldRegion = this._focusedRegion;

        if (oldRegion != null) {
            MessageLoop.sendMessage(oldRegion, DashboardLayoutRegion.LayoutMsg.BeforeBlur);
        }

        MessageLoop.sendMessage(region, DashboardLayoutRegion.LayoutMsg.BeforeFocus);

        this._focusedRegion = region;
        this._onFocusedChanged.emit(region);

        MessageLoop.sendMessage(region, DashboardLayoutRegion.LayoutMsg.AfterFocus);

        if (oldRegion != null) {
            MessageLoop.sendMessage(oldRegion, DashboardLayoutRegion.LayoutMsg.AfterBlur);
        }
    }

    private setDirty() {
        this.root.setFresh();
        MessageLoop.sendMessage(this.root, Widget.Msg.FitRequest);
        MessageLoop.sendMessage(this.root, Widget.Msg.UpdateRequest);
        if (this._isDirty) return;
        this._isDirty = true;
        this._OnDirtySrc.next();
    }
}

export namespace LayoutManager {
    /**
     * An interface for some part provider.
     *
     * Parts are normally provided by the part manager, but this interface
     * allows the layout engine to work with things that aren't quite parts.
     * The only requirement is that they be referencable by some string ID.
     *
     * @export
     * @interface IPartManager
     */
    export interface IPartManager {
        getPartById(id: string): Widget | null;
    }

    export interface IFactory {
        keys(): Iterable<string>;
        has(key: string): boolean;
        get(key: string): Widget | Promise<Widget>;
    }
}
