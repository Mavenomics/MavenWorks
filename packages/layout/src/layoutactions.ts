import { LayoutManager } from "./LayoutManager";
import { StackPanelLayoutRegion } from "./StackPanelLayoutRegion";
import { MessageLoop } from "@phosphor/messaging";
import { Widget } from "@phosphor/widgets";
import { RegionWithChildren } from "./RegionWithChildren";
import { TabPanelDashboardLayoutRegion } from "./TabPanelLayoutRegion";
import { DashboardLayoutRegion } from "./DashboardLayoutRegion";
import { CanvasLayoutRegion } from "./CanvasLayoutRegion";

export namespace LayoutActions {
    // TODO: Some sort of layout region factory
    export const enum ContainerTypes {
        StackPanel = "Stack Panel",
        TabPanel = "Tab Panel",
        CanvasPanel = "Canvas",
    }

    interface SurroundWithArgs {
        layoutManager: LayoutManager;
        regionType: ContainerTypes;
        target: string;
        prunable: boolean;
    }

    /**
     * Surround a layout region in-place with a new parent.
     *
     * @param layoutManager The manager controlling the layout.
     * @param regionType The type of region to add.
     * @param target The ID of the region to surround.
     *
     * @throws If the target ID does not exist in the layout.
     * @throws If the target region doesn't have a parent region.
     *
     * #### Notes
     *
     * This won't check to see if the parent is already of the given type.
     * That is, it will *always* surround with the given region. Callers
     * that use this function automatically should first check to see that
     * the parent is of the type they need.
     */
    export function SurroundWith(
        {layoutManager, regionType, target, prunable}: SurroundWithArgs
    ) {
        const region = layoutManager.getRegion(target);
        if (region == null) {
            throw Error("Target region not found in layout");
        }
        const parentRegion = region.parentRegion;
        if (parentRegion == null) {
            // This is because it's either the root, or not attached
            throw Error("Target region does not have a valid parent");
        }
        const index = parentRegion.widgets.indexOf(region);
        let newParent: RegionWithChildren<any>;
        switch (regionType) {
            case "Stack Panel":
                newParent = new StackPanelLayoutRegion(layoutManager);
                break;
            case "Tab Panel":
                newParent = new TabPanelDashboardLayoutRegion(layoutManager);
                break;
            case "Canvas":
                newParent = new CanvasLayoutRegion(layoutManager);
                break;
            default:
                throw Error("Unknown region type " + regionType);
        }
        parentRegion.copyAttachedProperties(region, newParent);
        newParent.setLayoutProperty("prunable", prunable);
        newParent.setFresh();
        // addChild will unhook it from the previous parent
        newParent.addChild(region);
        parentRegion.insertChild(index, newParent);
        region.activate();
        return newParent.id;
    }

    /**
     * 4 drop zones are possible on the root node.
     * These are along the 4 sides of the dashboard- dropping a node here will
     * cause all the current root nodes to be moved to a new container. That
     * new container will share half the root size with the target region to
     * drop.
     */
    export const enum RootDropZone {
        FarOuterLeft = "far-outer-left",
        FarOuterRight = "far-outer-right",
        FarOuterTop = "far-outer-top",
        FarOuterBottom = "far-outer-bottom",
    }

    interface MoveToRootArgs {
        layoutManager: LayoutManager;
        zone: RootDropZone;
        target: string;
    }

    /** Move a region to the layout root
     *
     * @param layoutManager The manager controlling the Layout
     * @param zone The drop zone to move the region to
     * @param target The ID of the region to move
     *
     * @throws if no region in the manager can be found with id `target`
     */
    export function MoveToRootZone(
        {layoutManager, zone, target}: MoveToRootArgs
    ) {
        const toMove = layoutManager.getRegion(target);
        if (toMove == null) {
            throw Error("Drag target invalid! No region with id " + target);
        }

        const rootChildren = Array.from(layoutManager.root.widgets);
        const newContainer = new StackPanelLayoutRegion(layoutManager);
        // create a new reference
        newContainer.properties = {...layoutManager.root.properties};
        newContainer.setLayoutProperty("prunable", true);
        newContainer.setStale();
        toMove.setLayoutProperty("flexSize", 1);
        for (const child of rootChildren) {
            newContainer.addChild(child);
        }
        switch (zone) {
            case RootDropZone.FarOuterBottom:
            case RootDropZone.FarOuterRight:
                layoutManager.root.addChild(newContainer);
                layoutManager.root.addChild(toMove);
                break;
            case RootDropZone.FarOuterTop:
            case RootDropZone.FarOuterLeft:
                layoutManager.root.addChild(toMove);
                layoutManager.root.addChild(newContainer);
                break;
        }
        const isRootHorizontal = (zone === RootDropZone.FarOuterLeft ||
                                  zone === RootDropZone.FarOuterRight);
        layoutManager.root.setLayoutProperty("horizontal", isRootHorizontal);
        layoutManager.root.pruneSubtree();
        MessageLoop.sendMessage(layoutManager, Widget.Msg.UpdateRequest);
    }

    /**
     * 9 drop actions are possible relative to a given reference region.
     *
     * A Center drop will create a tab panel and add both the reference and the
     * target to it. If the reference is already a tab panel, the target is
     * simply added to the reference.
     *
     * An Inner drop will create a stack panel and add both the reference and
     * the target to it. If the reference is a stack panel, and it's
     * \[\[horizontal\]\] property aligns with the direction of the drop (eg,
     * a vertical stack panel for inner-top or inner-bottom), then the target
     * is added to the reference.
     *
     * An Outer drop does the same thing as an inner drop, with the exception
     * that it works on the parent of the reference region.
     */
    export const enum RelativeDropZone {
        Center = "center",
        InnerLeft = "inner-left",
        InnerRight = "inner-right",
        InnerTop = "inner-top",
        InnerBottom = "inner-bottom",
        OuterLeft = "outer-left",
        OuterRight = "outer-right",
        OuterTop = "outer-top",
        OuterBottom = "outer-bottom",
    }

    interface MoveToZoneArgs {
        layoutManager: LayoutManager;
        zone: RelativeDropZone;
        target: string;
        reference: string;
    }

    /**
     * Move a region, relative to another region.
     *
     * @param layoutManager The manager controlling the layout
     * @param zone The drop zone to move the region to
     * @param target The ID of the region to move
     * @param reference The ID of the region that will be used as a reference
     *
     * @throws if `target` or `reference` aren't valid IDs in the layout
     * @throws if dropping to an outer target and the reference has no parent
     */
    export function MoveToRelativeZone(
        {layoutManager, zone, target, reference}: MoveToZoneArgs
    ) {
        const targetRegion = layoutManager.getRegion(target);
        let referenceRegion = layoutManager.getRegion(reference);

        if (targetRegion == null) {
            throw Error("Region to move not found in layout");
        }
        if (referenceRegion == null) {
            throw Error("Reference for relative move not found in layout");
        }

        switch (zone) {
            case RelativeDropZone.Center:
                let tabPanel: TabPanelDashboardLayoutRegion;
                if (referenceRegion instanceof TabPanelDashboardLayoutRegion) {
                    tabPanel = referenceRegion;
                } else if (referenceRegion.parentRegion instanceof TabPanelDashboardLayoutRegion) {
                    tabPanel = referenceRegion.parentRegion;
                } else {
                    tabPanel = layoutManager.getRegion(SurroundWith({
                        layoutManager,
                        target: reference,
                        regionType: ContainerTypes.TabPanel,
                        prunable: true
                    })) as TabPanelDashboardLayoutRegion;
                }
                tabPanel.addChild(targetRegion);
                tabPanel.setLayoutProperty("ForegroundIndex", tabPanel.widgets.length - 1);
                break;
            default:
                const stackPanel = layoutManager.getRegion(WrapIfNot(
                    StackPanelLayoutRegion,
                    [["horizontal", zone.endsWith("left") || zone.endsWith("right")]],
                    layoutManager,
                    // if it's an outer drop, use the parent stackpanel
                    zone.startsWith("inner") ? referenceRegion : referenceRegion.parentRegion
                )) as StackPanelLayoutRegion;
                MoveToStackPanel(stackPanel, targetRegion, zone);
        }
        layoutManager.root.pruneSubtree();

    }

    /**
     * Conditionally surround a region if it is not already a container.
     *
     * If the target is an instance of the `RegionClass`, then it is considered
     * a match and the target is returned without modification.
     *
     * If `propsMustMatch` is provided, each [property, value] pair will be
     * compared with the target. If any target\[\[property\]\] !== value, then
     * target will not be considered a match.
     *
     * @param RegionClass The constructor for the container to optionally create
     * @param propsMustMatch A set of properties that must match in order to re-use the region
     * @param layoutManager The layout manager
     * @param target The target region to conditionally surround
     */
    function WrapIfNot(
        RegionClass: { new(...args: any[]): RegionWithChildren<any> },
        propsMustMatch: [string, unknown][],
        layoutManager: LayoutManager,
        target: DashboardLayoutRegion | null
    ) {
        if (target == null) {
            throw Error("Cannot wrap a non-existent panel");
        }
        if (target instanceof RegionClass) {
            let didPropsMatch = true;
            for (const [propName, propValue] of propsMustMatch) {
                if (target.getLayoutProperty(propName as any) !== propValue) {
                    didPropsMatch = false;
                    break;
                }
            }
            if (didPropsMatch) {
                return target.id;
            }
        }
        return SurroundWith({
            layoutManager,
            target: target.id,
            prunable: true,
            regionType: (RegionClass === StackPanelLayoutRegion ? ContainerTypes.StackPanel : ContainerTypes.TabPanel)
        });
    }

    function MoveToStackPanel(
        reference: StackPanelLayoutRegion,
        target: DashboardLayoutRegion,
        zone: RootDropZone | RelativeDropZone
    ) {
        if (zone.endsWith("right") || zone.endsWith("bottom")) {
            reference.addChild(target);
        } else {
            reference.insertChild(0, target);
        }
        const isRootHorizontal = (zone.endsWith("left") || zone.endsWith("right"));
        reference.setLayoutProperty("horizontal", isRootHorizontal);
        return reference;
    }
}
