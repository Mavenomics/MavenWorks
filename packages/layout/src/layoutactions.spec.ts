import * as coreutils from "@phosphor/coreutils";

// Mock the UUID generator to force consistent snapshots
(coreutils as any).UUID.uuid4 = jest.fn(() => "<generated>");

// JSDOM doesn't support insertAdjacentElement, but we're not looking at the DOM
// so we don't need to do anything
Element.prototype.insertAdjacentElement = jest.fn();

import { LayoutManager } from "./LayoutManager";
import { Widget } from "@phosphor/widgets";
import { LayoutActions } from "./layoutactions";
import { TabPanelDashboardLayoutRegion } from "./TabPanelLayoutRegion";
import { LayoutSerializer } from "./LayoutSerializer";

const layout = Object.freeze({
    "properties": {},
    "typeName": 0,
    "uuid": "root",
    "children": [{
        "properties": {},
        "typeName": 0,
        "uuid": "d1241f21-346c-46c2-9010-493493fa7ee5",
        "children": [{
                "properties": {
                    "backgroundColor": "red",
                    "prunable": false,
                },
                "typeName": 0,
                "uuid": "e8cb8a20-cd0b-4c53-ac5f-8137dffd95fc",
                "children": []
            }, {
                "properties": {
                    "backgroundColor": "blue",
                    "prunable": false,
                },
                "typeName": 0,
                "uuid": "7f4d05ff-6ccd-4bd7-aac5-93f18e131c6d",
                "children": []
            }
        ]
    }],
});

describe("Layout Actions", () => {
    const layoutManager = new LayoutManager({
        getPartById: () => {
            const widget = new Widget();
            widget.addClass("test_widget");
        }
    } as any, {} as any);

    beforeEach(() => {
        layoutManager.initLayout(layout);
    });

    describe("SurroundWith", () => {
        test("shouldn't allow surrounding root", () => {
            expect(() => LayoutActions.SurroundWith({
                layoutManager,
                target: layoutManager.root.id,
                regionType: LayoutActions.ContainerTypes.TabPanel,
                prunable: false
            })).toThrow();
        });

        test("should surround with containers", () => {
            const target = "e8cb8a20-cd0b-4c53-ac5f-8137dffd95fc";
            LayoutActions.SurroundWith({
                layoutManager,
                target,
                regionType: LayoutActions.ContainerTypes.TabPanel,
                prunable: false
            });
            const parent = layoutManager.getParentRegion(target);
            expect(parent).not.toBeNull();
            expect(parent!.widgets).toHaveLength(1);
            expect(parent).toBeInstanceOf(TabPanelDashboardLayoutRegion);
        });

        test("should set [[prunable]] on generated regions", () => {
            const target = "7f4d05ff-6ccd-4bd7-aac5-93f18e131c6d";
            LayoutActions.SurroundWith({
                layoutManager,
                target,
                regionType: LayoutActions.ContainerTypes.TabPanel,
                prunable: true
            });
            const parent = layoutManager.getParentRegion(target);
            expect(parent!.getLayoutProperty("prunable")).toBeTruthy();
        });
    });

    test("Move to Far Outer Zone", () => {
        const target = "e8cb8a20-cd0b-4c53-ac5f-8137dffd95fc";
        LayoutActions.MoveToRootZone({
            layoutManager,
            target,
            zone: LayoutActions.RootDropZone.FarOuterLeft
        });
        const serializedLayout = LayoutSerializer.toJson(layoutManager.root as any);
        expect(serializedLayout).toMatchSnapshot();
    });

    describe("Relative drop zones", () => {
        const target = "e8cb8a20-cd0b-4c53-ac5f-8137dffd95fc";
        const reference = "7f4d05ff-6ccd-4bd7-aac5-93f18e131c6d";

        test("Move to Center Drop Zone", () => {
            LayoutActions.MoveToRelativeZone({
                layoutManager,
                target,
                reference,
                zone: LayoutActions.RelativeDropZone.Center
            });
            const serializedLayout = LayoutSerializer.toJson(layoutManager.root as any);
            expect(serializedLayout).toMatchSnapshot();
        });

        test("Move to Inner Drop Zone", () => {
            LayoutActions.MoveToRelativeZone({
                layoutManager,
                target,
                reference,
                zone: LayoutActions.RelativeDropZone.InnerRight
            });
            const serializedLayout = LayoutSerializer.toJson(layoutManager.root as any);
            expect(serializedLayout).toMatchSnapshot();
        });

        test("Move to Outer Drop Zone", () => {
            LayoutActions.MoveToRelativeZone({
                layoutManager,
                target,
                reference,
                zone: LayoutActions.RelativeDropZone.OuterRight
            });
            const serializedLayout = LayoutSerializer.toJson(layoutManager.root as any);
            expect(serializedLayout).toMatchSnapshot();
        });
    });
});
