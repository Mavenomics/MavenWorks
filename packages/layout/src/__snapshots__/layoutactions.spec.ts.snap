// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Actions Move to Far Outer Zone 1`] = `
Object {
  "attachedProperties": Array [
    Object {
      "Expanded": null,
      "Fixed Size (px)": null,
      "Stretch": 1,
    },
    Object {
      "Expanded": null,
      "Fixed Size (px)": null,
      "Stretch": 1,
    },
  ],
  "children": Array [
    Object {
      "attachedProperties": Array [],
      "children": Array [],
      "properties": Object {
        "backgroundColor": "red",
        "flexSize": 1,
        "prunable": false,
      },
      "typeName": 0,
      "uuid": "e8cb8a20-cd0b-4c53-ac5f-8137dffd95fc",
    },
    Object {
      "attachedProperties": Array [],
      "children": Array [],
      "properties": Object {
        "backgroundColor": "blue",
        "prunable": false,
      },
      "typeName": 0,
      "uuid": "7f4d05ff-6ccd-4bd7-aac5-93f18e131c6d",
    },
  ],
  "properties": Object {
    "horizontal": true,
  },
  "typeName": 0,
  "uuid": "root",
}
`;

exports[`Layout Actions Relative drop zones Move to Center Drop Zone 1`] = `
Object {
  "attachedProperties": Array [
    Object {
      "Expanded": null,
      "Fixed Size (px)": null,
      "Stretch": 1,
    },
  ],
  "children": Array [
    Object {
      "attachedProperties": Array [
        Object {},
        Object {},
      ],
      "children": Array [
        Object {
          "attachedProperties": Array [],
          "children": Array [],
          "properties": Object {
            "backgroundColor": "blue",
            "prunable": false,
          },
          "typeName": 0,
          "uuid": "7f4d05ff-6ccd-4bd7-aac5-93f18e131c6d",
        },
        Object {
          "attachedProperties": Array [],
          "children": Array [],
          "properties": Object {
            "backgroundColor": "red",
            "prunable": false,
          },
          "typeName": 0,
          "uuid": "e8cb8a20-cd0b-4c53-ac5f-8137dffd95fc",
        },
      ],
      "properties": Object {
        "ForegroundIndex": 1,
        "caption": "Tab Panel",
        "prunable": true,
      },
      "typeName": 2,
      "uuid": "<generated>",
    },
  ],
  "properties": Object {},
  "typeName": 0,
  "uuid": "root",
}
`;

exports[`Layout Actions Relative drop zones Move to Inner Drop Zone 1`] = `
Object {
  "attachedProperties": Array [
    Object {
      "Expanded": null,
      "Fixed Size (px)": null,
      "Stretch": 1,
    },
  ],
  "children": Array [
    Object {
      "attachedProperties": Array [
        Object {
          "Expanded": null,
          "Fixed Size (px)": null,
          "Stretch": 1,
        },
        Object {
          "Expanded": null,
          "Fixed Size (px)": null,
          "Stretch": 1,
        },
      ],
      "children": Array [
        Object {
          "attachedProperties": Array [],
          "children": Array [],
          "properties": Object {
            "backgroundColor": "blue",
            "prunable": false,
          },
          "typeName": 0,
          "uuid": "7f4d05ff-6ccd-4bd7-aac5-93f18e131c6d",
        },
        Object {
          "attachedProperties": Array [],
          "children": Array [],
          "properties": Object {
            "backgroundColor": "red",
            "prunable": false,
          },
          "typeName": 0,
          "uuid": "e8cb8a20-cd0b-4c53-ac5f-8137dffd95fc",
        },
      ],
      "properties": Object {
        "caption": "Stack Panel",
        "horizontal": true,
        "prunable": true,
      },
      "typeName": 0,
      "uuid": "<generated>",
    },
  ],
  "properties": Object {},
  "typeName": 0,
  "uuid": "root",
}
`;

exports[`Layout Actions Relative drop zones Move to Outer Drop Zone 1`] = `
Object {
  "attachedProperties": Array [
    Object {
      "Expanded": null,
      "Fixed Size (px)": null,
      "Stretch": 1,
    },
  ],
  "children": Array [
    Object {
      "attachedProperties": Array [
        Object {
          "Expanded": null,
          "Fixed Size (px)": null,
          "Stretch": 1,
        },
        Object {
          "Expanded": null,
          "Fixed Size (px)": null,
          "Stretch": 1,
        },
      ],
      "children": Array [
        Object {
          "attachedProperties": Array [],
          "children": Array [],
          "properties": Object {
            "backgroundColor": "blue",
            "prunable": false,
          },
          "typeName": 0,
          "uuid": "7f4d05ff-6ccd-4bd7-aac5-93f18e131c6d",
        },
        Object {
          "attachedProperties": Array [],
          "children": Array [],
          "properties": Object {
            "backgroundColor": "red",
            "prunable": false,
          },
          "typeName": 0,
          "uuid": "e8cb8a20-cd0b-4c53-ac5f-8137dffd95fc",
        },
      ],
      "properties": Object {
        "caption": "Stack Panel",
        "horizontal": true,
        "prunable": true,
      },
      "typeName": 0,
      "uuid": "<generated>",
    },
  ],
  "properties": Object {},
  "typeName": 0,
  "uuid": "root",
}
`;
